src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,282 @@
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newMemoryFromGenesisIO ledgerEval genesisBlk initialState =
+    let blks = Chrono.NewestFirst $ genesisBlk NonEmpty.:| []
+    in newMemoryIO ledgerEval blks initialState
+
+
+-- | Create a ledger backed by the given block store and in-memory
+-- storage of state and receipts.
+--
+-- @initialState@ is the state that the transactions of the
+-- genesisblock are applied to.
+newFromBlockStore
+    :: ( MonadIO m
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto tx
+       )
+    => Evaluator state tx output
+    -> BlockStore crypto tx seal IO
+    -> state
+    -> IO (Ledger crypto seal tx output state m)
+newFromBlockStore ledgerEval blockStore initialState = do
+    receiptStore <- ReceiptStore.newReceiptStoreIO
+    stateStore <- newHashStoreIO
+    storeHashContent stateStore initialState
+
+    let ledger = Ledger{blockStore, stateStore, receiptStore, ledgerEval}
+    materialize ledger initialState
+    pure $ hoist liftIO ledger
+
+
+mkLedger
+    :: BlockStore crypto tx seal m
+    -> HashStore crypto state m
+    -- ^ State store
+    -> Evaluator state tx output
+    -> ReceiptStore.ReceiptStore crypto tx output m
+    -> Ledger crypto seal tx output state m
+mkLedger blockStore stateStore ledgerEval receiptStore =
+    Ledger
+        { blockStore
+        , ledgerEval
+        , stateStore
+        , receiptStore
+        }
+
+
+-----------------------------------------------------------
+-- * Getters
+-----------------------------------------------------------
+
+blockStoreReader :: Ledger crypto seal tx output state m -> BlockStore.BlockStoreReader crypto tx seal m
+blockStoreReader = fst . blockStore
+
+-- | Get the @depth@ most recent blocks.
+getBlocksByDepth
+    :: Ledger crypto seal tx output state m
+    -> Depth
+    -> m (Chrono.NewestFirst [] (Block crypto tx (Sealed crypto seal)))
+getBlocksByDepth ledger depth = BlockStore.getBlocksByDepth (blockStoreReader ledger) depth
+
+getTip :: Ledger crypto seal tx output state m -> m (Block crypto tx (Sealed crypto seal))
+getTip ledger = BlockStore.getTip (fst $ blockStore ledger)
+
+
+-- | Get the receipt for a given transaction contained in the
+-- blockchain.
+--
+-- Returns 'Nothing' if the transaction is not contained in the
+-- blockchain or the receipts was not properly build.
+getReceipt
+    :: (Monad m)
+    => Ledger crypto seal tx output state m
+    -> Crypto.Hashed crypto tx
+    -> m (Maybe (ReceiptStore.Receipt crypto tx output))
+getReceipt ledger txHash = runMaybeT $ do
+    -- Ensure that the transaction is in the block chain. Just calling
+    -- 'ReceiptStore.lookupReceipt' is not sufficient since
+    -- 'buildNextBlock' adds receipts without the block being part of
+    -- the blockchain.
+    _ <- MaybeT $ BlockStore.lookupTx (blockStoreReader ledger) txHash
+    MaybeT $ ReceiptStore.lookupReceipt (receiptStore ledger) txHash
```

Comment:
I'm not sure I grok this function:
1) By "in the blockchain", do you mean the best chain?
2) Why do we care if it's in the blockchain or not? Why not return it if it exists in the store?

Reply:
> By "in the blockchain", do you mean the best chain?

The block store only stores the best chain so it does not make sense to talk about the “best chain” in that context.

> Why do we care if it's in the blockchain or not? Why not return it if it exists in the store?

The receipt store acts as a cache. As explained in the comment `buildNextBlock` will add receipts to the receipt store but this does not mean that the tx is accepted into the blockchain.